hydra-cardano-api/src/Hydra/Cardano/Api/TxOut.hs

```diff
@@ -66,19 +79,54 @@ findTxOutByScript utxo script =
  where
   version = plutusScriptVersion (proxyToAsType $ Proxy @lang)
   matchScript = \case
-    (_, TxOut (AddressInEra _ (ShelleyAddress _ (Ledger.ScriptHashObj scriptHash') _)) _ _) ->
+    (_, TxOut (AddressInEra _ (ShelleyAddress _ (Ledger.ScriptHashObj scriptHash') _)) _ _ _) ->
       let scriptHash = toShelleyScriptHash $ hashScript $ PlutusScript version script
        in scriptHash == scriptHash'
     _ ->
       False
 -- * Type Conversions
 
--- | Convert a cardano-api's 'TxOut' into a cardano-ledger 'TxOut'
+-- | Convert a cardano-ledger 'TxOut' into a cardano-api 'TxOut'
+fromLedgerTxOut :: Ledger.TxOut (ShelleyLedgerEra Era) -> TxOut ctx Era
+fromLedgerTxOut =
+  fromShelleyTxOut shelleyBasedEra
+
+-- | Convert a cardano-api 'TxOut' into a cardano-ledger 'TxOut'
 toLedgerTxOut :: TxOut CtxUTxO Era -> Ledger.TxOut (ShelleyLedgerEra Era)
 toLedgerTxOut =
   toShelleyTxOut shelleyBasedEra
 
--- | Convert a cardano-ledger's 'TxOut' into a cardano-api 'TxOut'
-fromLedgerTxOut :: Ledger.TxOut (ShelleyLedgerEra Era) -> TxOut ctx Era
-fromLedgerTxOut =
-  fromShelleyTxOut shelleyBasedEra
+-- | Convert a plutus 'TxOut' into a cardano-api 'TxOut'.
+-- NOTE: Reference scripts are not resolvable right now.
+-- NOTE: Requires the 'Network' discriminator (Testnet or Mainnet) because
+-- Plutus addresses are stripped off it.
+fromPlutusTxOut ::
+  (HasMultiAsset era, HasScriptData era, HasInlineDatums era, IsShelleyBasedEra era) =>
+  Network ->
+  Plutus.TxOut ->
+  TxOut CtxUTxO era
+fromPlutusTxOut network out =
+  TxOut addressInEra value datum ReferenceScriptNone
+ where
+  addressInEra = fromPlutusAddress network plutusAddress
+
+  value = TxOutValue multiAssetSupportedInEra $ fromPlutusValue plutusValue
+
+  datum = case plutusDatum of
+    NoOutputDatum -> TxOutDatumNone
+    OutputDatumHash (Plutus.DatumHash hashBytes) ->
+      TxOutDatumHash scriptDataSupportedInEra . unsafeScriptDataHashFromBytes $ fromBuiltin hashBytes
+    OutputDatum (Plutus.Datum datumData) ->
+      TxOutDatumInline inlineDatumsSupportedInEra $ toScriptData datumData
+
+  Plutus.TxOut plutusAddress plutusValue plutusDatum _ = out
+
+-- | Convert a cardano-api 'TxOut' into a plutus 'TxOut'. Returns 'Nothing'
+-- if a byron address is used in the given 'TxOut'.
+toPlutusTxOut :: TxOut CtxUTxO Era -> Maybe Plutus.TxOut
+toPlutusTxOut =
+  -- NOTE: Since recently, the txInfoOutV2 conversion does take this
+  -- 'TxOutSource' to report corresponding 'TranslationError'. However, this
+  -- value is NOT used for constructing the Plutus.TxOut and hence we hard-code
+  -- it to 0 to not need to bookkeep it and we throw away errors anyway.
+  eitherToMaybe . Ledger.txInfoOutV2 (TxOutFromOutput $ Ledger.TxIx 0) . toLedgerTxOut
```

Comment:
Instead of hard-coding a TxIx, we could use a `error "TxOutSource" used but shouldn't` was we do not expect this to ever be used / evaluated (because of `eitherToMaybe`)